### PR TITLE
[16] Add lightgbm model for predicting company level future investment

### LINKED ISSUE #16: Add lightgbm model for predicting company level future investment

### PR DESCRIPTION
**Pull request by Jack-0-0**
Closes #16.

This PR adds:
* Notebook for training a model that can predict company future success. 
* Notebook for using the model to create an explainer dashboard.

Before running the notebooks, run `make inputs-pull` to get:
* `inputs/data/company_level/company_data_window_2011-01-01-2019-01-01.csv` -- dataset for training the lightgbm model
* `outputs/models/light_gbm_21h25m19s_2022-04-27` -- an example of the outputs from training the lightgbm model which can then be used for the explainer dashboard

### CODE REVIEW

**Comment by beingkk** [comment]
Hi Jack, thanks so much!

### Issue with the make file

When I was running `make inputs-pull` it returned the following error:
```shell
(iss_forecasting) ➜  iss_forecasting git:(16_light) make inputs-pull              
source .cookiecutter/scripts/import.sh && conda_activate iss_forecasting &&  aws s3 sync s3:///inputs inputs
fatal error: Parameter validation failed:
Invalid bucket name "": Bucket name must match the regex "^[a-zA-Z0-9.\-_]{1,255}$" or be an ARN matching the regex "^arn:(aws).*:(s3|s3-object-lambda):[a-z\-0-9]*:[0-9]{12}:accesspoint[/:][a-zA-Z0-9\-.]{1,63}$|^arn:(aws).*:s3-outposts:[a-z\-0-9]+:[0-9]{12}:outpost[/:][a-zA-Z0-9\-]{1,63}[/:]accesspoint[/:][a-zA-Z0-9\-]{1,63}$"
make: *** [inputs-pull] Error 1
```

The pulling does work as expected when I run: 

```
aws s3 sync s3://iss-forecasting/inputs inputs
```

Looks like the make file doesn't now the name of the s3 bucket - perhaps it's missing a line, something like
```shell
include .envrc
```
Alternatively, I guess you could define the bucket name in `.cookiecutter/config` ? 

Also, NB: Pulling inputs won't pull the model. For this I ran the command below. Perhaps we could have instructions to this effect in the repo readme?
```
aws s3 sync s3://iss-forecasting/outputs/models outputs/models
```

### Training

After pulling the inputs and model using the aws cli directly, I ran the two notebooks on my laptop and it overall worked well!

**Training time:** It would be useful to write down somewhere an estimate how long it takes to run the notebook (I guess 1hr for parameter search, a few seconds for one training iteration).

**Parameter intuition:** I was wondering if for the blog post it makes sense to provide a bit of intuition behind the parameters and values that you use, especially the ones for the grid search ([this article](https://neptune.ai/blog/lightgbm-parameters-guide) has some nice explanations). For example:
```python
    "num_leaves": [32, 35, 48],
    "max_depth": [11, 12, 13],
    "scale_pos_weight": [2.5, 2.55],
    "min_child_weight": [0.002, 0.003],
    "subsample": [0.4, 1],
    "colsample_bytree": [0.81, 0.83],
```
And perhaps your reflection, if possible, on the parameters that it found to best performing? Do they tell us anything about the data...?

Also curious, did you try any other values for the `"boosting"` parameter? eg, I guess `dart` might increase generalisation because it has dropout? I gave it a quick try with the best parameters from the parameter search when using `"boosting"=` and and have pushed an update - it takes quite long to train though (still training as I'm writing this :/ )

### Explainer dashboard

**Unpickling the model**: For the explainer notebook, it complained about the model pickle file:

```
Trying to unpickle estimator LabelEncoder from version 0.24.2 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
```

It does seem to work without issues. Do you think this could be a problem later on? Is there a way to use the appropriate pickle version for unpickling models...? Or is the best course of action perhaps simply to retrain the model with the best parameters, save with whatever pickle version I have, and use that for the dashboard?

**Dashboard vs notebook:** The dashboard's confusion matrix for the model you've uploaded on s3 looks a bit different than the confusion matrix that I ended up with when re-training the model using the notebook. I guess the exact values depend on the threshold probability? What's the threshold prob for the notebook's figure?

Notebook:
Screenshot 2022-06-24 at 15 04 10

Dashboard:
Screenshot 2022-06-24 at 15 04 19

**Blog/debrief session:** I think it would be great to highlight some of the most interesting plots from the explainability dashboard in the blog and/or debrief session. In particular I think these are quite interesting. 

newplot

I think this one (above) has quite a hopeful message, that when the prediction prob is particularly high (eg, larger than 0.7) the chance for actual success might be high. This suggests that it's not a bad idea to check the companies at the top of the list.

Screenshot 2022-06-24 at 15 34 19

For this one (above) you could even use screencastify to make gif for the blog or slides if you want to show mouse hovers... By the way, interestingly in that specific example (Ventive, a heating tech company) the feature "has_received_ukri_grant=1" increases the prediction by ~5% but the "n_grants=8" decreases by ~1%. Go figure (too many grants = not great?).

By the way, regarding SHAP values and their interpretability, [this looks like a fascinating paper](https://arxiv.org/pdf/2002.11097.pdf) (especially the last section).

**Missing companies:** I tried to find a couple companies in the dashboard's drop down menu, such as Mixergy or Sunamp (promising low carbon heating companies) but couldn't find them. What where the criteria again for excluding companies?

Best wishes,
Karlis

**Review by beingkk — APPROVED** [review]
Hey, thanks so much - this is great! In principle, I'm happy for you to merge this. I've just added some more general comments and questions, some of which relate to the blog and/or debrief session.

**Comment by Jack-0-0** [comment]
> Hi Jack, thanks so much!
> 
> ### Issue with the make file
> When I was running `make inputs-pull` it returned the following error:
> 
> ```shell
> (iss_forecasting) ➜  iss_forecasting git:(16_light) make inputs-pull              
> source .cookiecutter/scripts/import.sh && conda_activate iss_forecasting &&  aws s3 sync s3:///inputs inputs
> fatal error: Parameter validation failed:
> Invalid bucket name "": Bucket name must match the regex "^[a-zA-Z0-9.\-_]{1,255}$" or be an ARN matching the regex "^arn:(aws).*:(s3|s3-object-lambda):[a-z\-0-9]*:[0-9]{12}:accesspoint[/:][a-zA-Z0-9\-.]{1,63}$|^arn:(aws).*:s3-outposts:[a-z\-0-9]+:[0-9]{12}:outpost[/:][a-zA-Z0-9\-]{1,63}[/:]accesspoint[/:][a-zA-Z0-9\-]{1,63}$"
> make: *** [inputs-pull] Error 1
> ```
> 
> The pulling does work as expected when I run:
> 
> ```
> aws s3 sync s3://iss-forecasting/inputs inputs
> ```
> 
> Looks like the make file doesn't now the name of the s3 bucket - perhaps it's missing a line, something like
> 
> ```shell
> include .envrc
> ```
> 
> Alternatively, I guess you could define the bucket name in `.cookiecutter/config` ?
> 
> Also, NB: Pulling inputs won't pull the model. For this I ran the command below. Perhaps we could have instructions to this effect in the repo readme?
> 
> ```
> aws s3 sync s3://iss-forecasting/outputs/models outputs/models
> ```

Thank you for the review @beingkk.

Re `make inputs-pull`, I'm not sure what the issue is here. I'll make an issue for it and look into it.

I've added instructions in the readme on how to download the outputs.

**Comment by Jack-0-0** [comment]
> ### Training
> After pulling the inputs and model using the aws cli directly, I ran the two notebooks on my laptop and it overall worked well!
> 
> **Training time:** It would be useful to write down somewhere an estimate how long it takes to run the notebook (I guess 1hr for parameter search, a few seconds for one training iteration).

I've added time estimates to the notebook.

> **Parameter intuition:** I was wondering if for the blog post it makes sense to provide a bit of intuition behind the parameters and values that you use, especially the ones for the grid search ([this article](https://neptune.ai/blog/lightgbm-parameters-guide) has some nice explanations). For example:
> 
> ```python
>     "num_leaves": [32, 35, 48],
>     "max_depth": [11, 12, 13],
>     "scale_pos_weight": [2.5, 2.55],
>     "min_child_weight": [0.002, 0.003],
>     "subsample": [0.4, 1],
>     "colsample_bytree": [0.81, 0.83],
> ```
> 
> And perhaps your reflection, if possible, on the parameters that it found to best performing? Do they tell us anything about the data...?
>

I've mentioned using hyperparameter search in the blog. Most of the parameters do not make a huge difference to the performance. The biggest impact on performance was `scale_pos_weight` so I've talked about that.


> Also curious, did you try any other values for the `"boosting"` parameter? eg, I guess `dart` might increase generalisation because it has dropout? I gave it a quick try with the best parameters from the parameter search when using `"boosting"=` and and have pushed an update - it takes quite long to train though (still training as I'm writing this :/ )

I've tried using `dart` but could not achieve as good performance as using `gbdt`. `dart` is more difficult to use with automated hyperparameter search as you cannot use `early stopping` with it.

**Comment by Jack-0-0** [comment]
> ### Explainer dashboard
> **Unpickling the model**: For the explainer notebook, it complained about the model pickle file:
> 
> ```
> Trying to unpickle estimator LabelEncoder from version 0.24.2 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
> https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
> ```
> 
> It does seem to work without issues. Do you think this could be a problem later on? Is there a way to use the appropriate pickle version for unpickling models...? Or is the best course of action perhaps simply to retrain the model with the best parameters, save with whatever pickle version I have, and use that for the dashboard?

This should be OK as the requirements.txt has the versions specified.

> 
> **Dashboard vs notebook:** The dashboard's confusion matrix for the model you've uploaded on s3 looks a bit different than the confusion matrix that I ended up with when re-training the model using the notebook. I guess the exact values depend on the threshold probability? What's the threshold prob for the notebook's figure?

I checked this and the confusion matrix matches for me in the notebook vs. the dashboard. The notebook uses 0.5 threshold and the dashboard defaults to 0.5. Are you looking at `models/light_gbm_21h25m19s_2022-04-27` for both?

**Comment by Jack-0-0** [comment]
> **Blog/debrief session:** I think it would be great to highlight some of the most interesting plots from the explainability dashboard in the blog and/or debrief session. In particular I think these are quite interesting.
>
> I think this one (above) has quite a hopeful message, that when the prediction prob is particularly high (eg, larger than 0.7) the chance for actual success might be high. This suggests that it's not a bad idea to check the companies at the top of the list.
> 
> For this one (above) you could even use screencastify to make gif for the blog or slides if you want to show mouse hovers... By the way, interestingly in that specific example (Ventive, a heating tech company) the feature "has_received_ukri_grant=1" increases the prediction by ~5% but the "n_grants=8" decreases by ~1%. Go figure (too many grants = not great?).

Thanks, I have added some of these charts to the blog now and added the SHAP dependence plot for `n_grants`.

> By the way, regarding SHAP values and their interpretability, [this looks like a fascinating paper](https://arxiv.org/pdf/2002.11097.pdf) (especially the last section).
> 

I'll take a look 🙂

> **Missing companies:** I tried to find a couple companies in the dashboard's drop down menu, such as Mixergy or Sunamp (promising low carbon heating companies) but couldn't find them. What where the criteria again for excluding companies?

The companies are younger than 8 years old and have not already been acquired or had an IPO.
Mixergy is in the dataset but I couldn't find Sunamp. The explainer dashboard is looking at the validation set, so if Mixergy was in the training set it won't be in the explainer dashboard.
